tests/python/integration/test_ewise_fpga.py

```diff
@@ -0,0 +1,46 @@
+import tvm
+import numpy as np
+import os
+
+os.environ["XCL_EMULATION_MODE"] = "1"
+
+@tvm.register_func
+def tvm_callback_vhls_postproc(code):
+    """Hook to inspect the Vivado HLS code before actually run it"""
+    print(code)
+    return code
+
+def test_exp():
+    # graph
+    n = tvm.convert(1024)
+    A = tvm.placeholder((n,), name='A')
+    B = tvm.compute(A.shape, lambda *i: tvm.exp(A(*i)), name='B')
+    s = tvm.create_schedule(B.op)
+    # create iter var and assign them tags.
+    px, x = s[B].split(B.op.axis[0], nparts=1)
+    s[B].bind(px, tvm.thread_axis("pipeline"))
+
+    # one line to build the function.
+    def check_device(device, host="stackvm"):
+        if not tvm.module.enabled(host):
+            return
+        ctx = tvm.context(device, 0)
+        if not ctx.exist:
+            return
+        fexp = tvm.build(s, [A, B],
+                         device, host,
+                         name="myexp")
+        ctx = tvm.context(device, 0)
+        # launch the kernel.
+        n = 1024
+        a = tvm.nd.array(np.random.uniform(size=n).astype(A.dtype), ctx)
+        b = tvm.nd.array(np.zeros(n, dtype=B.dtype), ctx)
+        fexp(a, b)
+        np.testing.assert_allclose(
+            b.asnumpy(), np.exp(a.asnumpy()), rtol=1e-5)
+
+    check_device("sdaccel")
```

Comment:
Does the TVM Jenkins environment has sdaccel?
If so, how did you mock the bitstream generation part?
If not, it seems to me that this test will exit at line 29 and nothing will be tested.

Reply:
I think the Jenkins environment of this community doesn't have SDAccel, so the test is not run via CI.  I checked the test works on my AWS F1 instance:
```
$ python -m nose -vs tests/python/integration/test_ewise_fpga.py
test_ewise_fpga.test_exp ... [07:37:36] /home/centos/src/project_data/tvm/src/runtime/opencl/opencl_device_api.cc:232: Initialize OpenCL platform 'Xilinx'
[07:37:36] /home/centos/src/project_data/tvm/src/runtime/opencl/opencl_device_api.cc:259: opencl(0)='xilinx_aws-vu9p-f1_dynamic_5_0' cl_device_id=0x1ad7ef0
#include <ap_int.h>

extern "C" void myexp_kernel0( float* B,  float* A) {
#pragma HLS INTERFACE m_axi port=B  offset=slave bundle=gmem
#pragma HLS INTERFACE s_axilite port=B bundle=control
#pragma HLS INTERFACE m_axi port=A  offset=slave bundle=gmem
#pragma HLS INTERFACE s_axilite port=A bundle=control
#pragma HLS INTERFACE s_axilite port=return bundle=control

  for (int i0_inner = 0; i0_inner < 1024; ++i0_inner) {
    B[i0_inner] = expf(A[i0_inner]);
  }
}



****** xocc v2017.4 (64-bit)
  **** SW Build 2193837 on Tue Apr 10 18:06:59 MDT 2018
    ** Copyright 1986-2017 Xilinx, Inc. All Rights Reserved.

Attempting to get a license: ap_opencl
Feature available: ap_opencl
INFO: [XOCC 60-585] Compiling for software emulation target

(snip)

INFO: [XOCC 60-586] Created /tmp/tmpc5uk9a/output.xclbin
INFO: [XOCC 60-791] Total elapsed time: 0h 0m 11s
ok

----------------------------------------------------------------------
Ran 1 test in 61.040s

OK
```

I set the environment variable XCL_EMULATION_MODE in the python test file, so SDAccel doesn't generate a bitstream but run in software emulation mode - it works like a mock.

Let me know if I'm missing something.

Reply:
Another way is to use FPGA Developer AMI of AWS.  It contains SDAccel.

@tqchen, is it possible to launch an AWS instance from Jenkins for FPGA testing purpose?  Note that we don't need F1 instances for emulation.  Any instance types is okay.